Module tabulation

cgarciae opened the pull request:
# What does this PR do?
Adds mechanism to tabulate a Module, similar to Haiku's `tabulate` or Kera's `summary`.  This builds on the ideas/comments from #288 and aims to use `rich` to ultimately generate a nice table representation of the module. As previously discussed, the idea is to implement 2 steps:

1. Generate an intermediate representation
2. Render such representation

The PR has the following goals:

- [x] Create function to extract an intermediate representation for the Module (e.g. `module_info`).
- [x] Create a function that takes the intermediate representation and render is as a string (e.g. `render_table`).
- [x] Add a method to Module that directly render the module to a string (e.g. `Module.tabulate`).
- [x] Add docs
- [x] Add tests

**NOTE**: current names used are open for debate.

cgarciae commented:
@avital @marcvanzee 

The proposed `module_info` function takes in a Module and sample inputs and returns a `Table` which is just a list of `Row`s, where Row is:

```python
@dataclass
class Row:
  path: Tuple[str, ...]
  outputs: Any # outputs of the module, shapes and dtypes only
  module_variables: Dict[str, Dict[str, Array]] # module's immediate variables only 
  counted_variables: Dict[str, Dict[str, Array]] # all variables that should be counted for this row
```
The idea is that with this information you can later create a meaningful representation. Under the hood its implemented using `jax.eval_shape` over `Module.init` with `capture_intermediate=True`. It has a `depth` argument which can be used to abbreviate the table e.g. `depth=1` will stop the recursion after 1 level so it will only show the top Module and its immediate submodules. `counted_variables` specifies which variables should be counted for that row (e.g. total number or total bytes), in most cases it will be the same as `module_variables` except for when max depth is reached in which case `counted_variables` contains the variables for that module and all of its sub-modules.

cgarciae commented:
Next I'll implement a `render_table` function that uses rich to create a nice table representation. Design-wise it will look somewhat like this but without the `module` and `inputs` columns:
![150032598-3e39f1a4-1a88-4303-9a95-6a8784a8b894](https://user-images.githubusercontent.com/5862228/152050716-b638b1fc-cb51-4bbb-ab02-7ce5c2423b9c.png)

The last columns are going to be variable depending on the collections present e.g `Rng` will be removed and `Parameter` and `BatchStat` will be replace with `params` and `batch_stats`.

cgarciae commented:
Gave it a go at an initial implementation of `render_table`, here is a sample of how it currently looks:

![Screenshot from 2022-02-02 09-18-38](https://user-images.githubusercontent.com/5862228/152171373-50f697a3-1f4b-4739-9f00-6909cc853404.png)

Some notes:
* `variables` and `outputs` are formatted with `yaml`.
* arrays are represented as `{dtype}[{shape}]`, similar how JAX represents tracers.
* need to add checks to avoid empty collections being shown

Feedback on design appreciated.

cgarciae commented:
@avital 

Even though `ConvBlock` is defined like this:

```python
@dataclasses.dataclass
class ConvBlock(nn.Module):
  features: int
  kernel_size: List[int]
  training: bool

  @nn.compact
  def __call__(self, x: Array) -> Array:
      x = nn.Conv(self.features, self.kernel_size)(x)
      x = nn.BatchNorm(use_running_average=not self.training)(x)
      x = nn.Dropout(0.5, deterministic=not self.training)(x)
      x = nn.relu(x)
      return x
```
Since the `variables` keys are alphabetically ordered e.g:

```
    intermediates: {
        ConvBlock_0: {
            BatchNorm_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
            },
            Conv_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
            },
            Dropout_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
            },
            __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 32), dtype=float32),),
        },
        ConvBlock_1: {
            BatchNorm_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
            },
            Conv_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
            },
            Dropout_0: {
                __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
            },
            __call__: (ShapeDtypeStruct(shape=(32, 28, 28, 64), dtype=float32),),
        },
        Dense_0: {
            __call__: (ShapeDtypeStruct(shape=(32, 10), dtype=float32),),
        },
        __call__: ((ShapeDtypeStruct(shape=(32, 10), dtype=float32), {'a': ShapeDtypeStruct(shape=(32, 10), dtype=float32), 'b': ShapeDtypeStruct(shape=(32, 10), dtype=float32)}),),
```
It would appear as if `BatchNorm` happened before `Conv` which can be confusing if you don't know about this.

cgarciae commented:
I thought that it would be nice if we showed the shapes of the inputs to the main module so I added a row with that. Here is how the latest version looks:

![Screenshot from 2022-02-03 08-31-48](https://user-images.githubusercontent.com/5862228/152352902-ba0df4b6-6afc-494c-9c2a-0ef7a570d64a.png)
 
 Here is the same Module but simplified with `depth=1`:
 
 
![Screenshot from 2022-02-03 08-32-18](https://user-images.githubusercontent.com/5862228/152352987-fa24ad39-a2b9-4623-b59a-06790233fe91.png)

avital commented:
This looks really cool! One question about the design -- why make separate columns for each variable collection? Seems potentially wasteful as these columns are always only one line, whereas the others may be multiple lines. Instead, maybe you could fit the parameter count and memory footprint directly within the column for variables, something like this maybe?

[...oh! you did the layout this way so that you could summarize the total at the bottom...]

So, maybe another proposal is to drop the `variables` column entirely and embed all of the information that's there into the respective column for each of the variable collections, e.g.:

```
batch_stats
---

64 (256 B)
  mean: float32[32]
  var: float32[32]
```

cgarciae commented:
@avital we could strive for removing the `variables` column, however, it might look strange when you use the `depth` parameter to get less rows. Currently the `variables` column shows the variables created by that module only (not the variables by its submodules) and columns like `params` or `batch_stats` shows the counts for that module and all its submodules when its submodules are filtered out by `depth`.

This is under the assumption we want to have the `depth` filter / summarization mechanism (do we?). I kinda like `depth` because when for big models it can give you a better overview if you set depth to low values like `1` or `2`. That said, this is very subjective an I am open to ideas.

cgarciae commented:
@avital I implemented your suggestion of removing the `variables` column and I quick like it:

![Screenshot from 2022-02-04 10-06-47](https://user-images.githubusercontent.com/5862228/152552588-6e24881c-9753-4fa9-afc9-2259509d42b6.png)

Here is with `depth=1`:

![Screenshot from 2022-02-04 10-07-44](https://user-images.githubusercontent.com/5862228/152552624-476cea4c-1c35-4715-bcd9-db3ac7a5ce75.png)

avital commented:
Hi @cgarciae and sorry for the late reply. I agree, I think this looks very nice!

cgarciae commented:
@avital I proceeded to create the `tabulate` method. Will be adding tests next.

Also, this PR introduced a lot of helper function and some classes, organization-wise are they ok there or should we move them?

avital commented:
@cgarciae I propose moving all of the new code to flax/linen/summary.py rather than module.py. I guess we should keep the `tabulate` method on Module for sake of discoverability but all of its implementation is probably best in a separate file. And I propose treating `module_info` (maybe rename `module_summary` in line with my proposed file name? Or we could unify on some different name) and `render_table` as public APIs as people may want to, e.g. run `module_info` on one machine and `render_table` on another, e.g. when serializing...

cgarciae commented:
@avital thanks for the feedback! 

* Moved helper functions/classes to `summary.py` as suggested
* Rename `module_info` to `module_summary`. 
* Exported `module_summary` and `render_table` as public attributes in the `linen` module.
* Added 3 tests.
* Added docs for `module_summary` and `render_table`.

cgarciae commented:
Hey @jheek! I had a talk with @marcvanzee, he suggested that I asked you if you could review this PR since both him and @avital are unavailable right now 🙏‍

marcvanzee commented:
@cgarciae apologies for the superlate reply. As you know we have been very busy and understaffed lately, but hopefully that will become better now! I will review your PR now. Can you please rebase and resolve the conflicts in the meantime?

cgarciae commented:
Hey @marcvanzee! Fixed merge issues.

Check the `test_tabulate` test if you want to play with it. One last pending issue to discuss here is that rows are not ordered by call order but alphabetically instead, in the future it would nice to modify the `capture_intermediates` feature to somehow indicate call order as well.

cc: @jheek

marcvanzee reviewed (changes requested):
First of all, this is going to be a really great addition! Users often struggle with printing their params and they always use their own little util functions.

As a high-level comment: I am having some problems rendering this table in the Python interpreter:

```
>>> nn.Dense(4).tabulate(random.PRNGKey(0), jnp.ones((3, 4)))
'\n\x1b[3m                 Dense Summary                  \x1b[0m\n┏━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┓\n┃\x1b[1m \x1b[0m\x1b[1mpath  \x1b[0m\x1b[1m \x1b[0m┃\x1b[1m \x1b[0m\x1b[1moutputs     \x1b[0m\x1b[1m \x1b[0m┃\x1b[1m \x1b[0m\x1b[1mparams              \x1b[0m\x1b[1m \x1b[0m┃\n┡━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━┩\n│ Inputs │ \x1b[2mfloat32\x1b[0m[3,4] │                      │\n├────────┼──────────────┼──────────────────────┤\n│ *      │ \x1b[2mfloat32\x1b[0m[3,4] │ bias: \x1b[2mfloat32\x1b[0m[4]     │\n│        │              │ kernel: \x1b[2mfloat32\x1b[0m[4,4] │\n│        │              │                      │\n│        │              │ \x1b[1m20 \x1b[0m\x1b[1;2m(80 B)\x1b[0m            │\n├────────┼──────────────┼──────────────────────┤\n│\x1b[1m \x1b[0m\x1b[1m      \x1b[0m\x1b[1m \x1b[0m│\x1b[1m \x1b[0m\x1b[1m       Total\x1b[0m\x1b[1m \x1b[0m│\x1b[1m \x1b[0m\x1b[1m20 \x1b[0m\x1b[1;2m(80 B)\x1b[0m\x1b[1m           \x1b[0m\x1b[1m \x1b[0m│\n└────────┴──────────────┴──────────────────────┘\n\x1b[1m                                                \x1b[0m\n\x1b[1m          Total Parameters: 20 \x1b[0m\x1b[1;2m(80 B)\x1b[0m\x1b[1m           \x1b[0m\n\n'
```

This happens both in my ITerm terminal and in my VSCode terminal.

I am using UTF-8 encoding and I have coloring enabled in Iterm. Any idea why this happens?

I suppose something is wrong with my setup, but given that I run into this now, I fear that it may happen to other users as well...

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,435 @@
+    self.collections = collections
+
+
+def module_summary(
```
The name seems to suggests this will return a summary, but it actually returns a Table object, which isn't very useful, so I wonder if it is useful to expose this function at all. 

Alternatively you could add a function `tabulate` to this file that calls `module_summary` (I would rename it to `_get_module_table`) and `render_table` (I would rename it to `_render_table`).

Otherwise, users may try to call `summary.module_summary` expecting a summary of the model.

cgarciae replied:
Plan:
- rename `module_summary` => `_get_module_table`
- rename `render_table` => `_render_table`
- create `summary.tabulate` that follows @jheek transformation style suggestion.

cgarciae commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,435 @@
+            return nn.Dense(2)(h)
+
+    x = jnp.ones((16, 9))
+    model = Foo()
+
+    table = nn.module_summary(model, jax.random.PRNGKey(0), x))
```
This is shorter but I am not sure its clearer 🤔

cgarciae replied:
Ok, will use the short version.

cgarciae commented on flax/linen/module.py:
```diff
@@ -1359,6 +1362,93 @@ def __call__(self, x):
+    rngs: Union[PRNGKey, RNGSequences],
+    *args,
+    method: Optional[Callable[..., Any]] = None,
+    mutable: CollectionFilter = DenyList("intermediates"),
```
Sounds good, I was aiming for `tabulate` to have (almost) the same signature as `init` so this was copied over, but it wouldn't hurt to have `tabulate` be as flexible as possible.

cgarciae commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,435 @@
+    self.collections = collections
+
+
+def module_summary(
```
I like this! Will do.

BTW: I wish `Module.init/apply` did this as well.

cgarciae commented on flax/linen/__init__.py:
```diff
@@ -35,6 +35,8 @@
 from .pooling import avg_pool, max_pool, pool
 from .recurrent import ConvLSTM, GRUCell, LSTMCell, OptimizedLSTMCell
 from .stochastic import Dropout
+from .initializers import ones, zeros
```
Doesn't register as a change from this PR 🤔

cgarciae commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,435 @@
+  if len(module_outputs) == 0:
+    output = None
+  elif len(module_outputs) > 1:
+    moduel_path = "/".join(path)
```
I'll rename it to `path_repr` which makes more sense.

cgarciae replied:
On second thoughts I just inlined it.

cgarciae commented on tests/linen/linen_module_test.py:
```diff
@@ -17,26 +17,25 @@
 import dataclasses
```
sounds good!

cgarciae commented on tests/linen/linen_module_test.py:
```diff
@@ -1382,6 +1381,170 @@ class MyComponent2(Generic[T], nn.Module):
+    )
+
+    # 11 rows = 1 Inputs + 4 ConvBlock_0 + 4 ConvBlock_1 + 1 Dense_0 + 1 Module output
+    assert len(table) == 11
```
Pytest somehow overrides `assert` such that if you change this `11` to a `12` the error message when you run this test like this:

```bash
pytest tests/linen/summary_test.py
```
is presented in a rich form as:
```
>     assert len(table) == 12
E     AssertionError: assert 11 == 12
E      +  where 11 = len([Row(path=('Inputs',), outputs=(DeviceArray([[[[1.],\n               [1.],\n               [1.],\n               ...,\n   ...2), {'a': ShapeDtypeStruct(shape=(32, 10), dtype=float32), 'b': ShapeDtypeStruct(shape=(32, 10), dtype=float32)}),)}})])
```
Using `assertEquals` looks like this:
```
>     self.assertEqual(len(table), 12)
E     AssertionError: 11 != 12
```

cgarciae replied:
Ok, I see `assertEqual` is used throughout the code base so I will change it to match the style.

cgarciae replied:
I do prefer the `assert` style as it looks more natural.

marcvanzee replied:
`self.assertEqual` provides a bit more functionality than just `assert`. It provides better error messages in the case of failure, and it will automatically call `assertListEqual`, `assertDictEqual`, ... if the two inputs have the same type.

cgarciae commented on flax/linen/module.py:
```diff
@@ -1353,6 +1355,93 @@ def __call__(self, x):
+        depth limit, its parameter count and bytes will be added to the row of 
+        its first shown ancestor such that the sum of all rows always adds up to 
+        the total number of parameters of the Module.
+      output_methods: Method names in the `intermediates` collection that should be
```
Added brief description for `*args` and `**kwargs`.

cgarciae commented:
Hey @marcvanzee and @jheek!
Fixed a couple issues based on your feedback. Its ready for another round 😀

jheek commented:
@cgarciae you can have at most 5 commits in a PR. Please squash / rebase to fix the build

jheek commented on flax/linen/module.py:
```diff
@@ -1353,6 +1355,95 @@ def __call__(self, x):
     self.scope.put_variable(col, name, xs)
     return True
 
+  def tabulate(
```
tabulate does not yet show up in the docs. You still need to add it to flax.linen.rst Note you can click "details" on the RTD action to see a preview of how the docs will look

jheek replied:
I think inspection/summarization makes sense

marcvanzee commented:
> I think inspection/summarization makes sense

+1 to inspection!

cgarciae commented:
Created the `Inspection` section.

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+  return size, bytes
+
+def _get_rich_repr(obj):
+  import io
+
+  import rich.console
```
Imports should be at the top.

cgarciae replied:
Done!

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+      show_lines=True,
+      show_footer=True,
+      title=f"{table.module.__class__.__name__} Summary",
+      # box=rich.box.HORIZONTALS,
```
Nit: Remove?

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+  )
+
+  rich_table.add_column("path")
+  # rich_table.add_column("variables")
```
Nit: Remove?

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+
+    rich_table.add_row(
+      "/".join(row.path) if row.path else table.module.__class__.__name__,
+      # _as_yaml_str(jax.tree_map(_format_value, row.module_variables)),
```
Nit: Remove?

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+  rich_table.columns[1].footer = rich.text.Text.from_markup("Total", justify="right")
+
+  # get collection totals
+  colletion_total = {col: (0, 0) for col in table.collections}
```
typo: `colletion_total` --> `collection_total`

cgarciae replied:
good catch!

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+  rich_table.caption = f"\nTotal Parameters: {_size_and_bytes_repr(caption_totals)}"
+
+  return "\n" + _get_rich_repr(rich_table) + "\n"
+
```
```suggestion
```

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+  return f"{size:,} [dim]({bytes_repr})[/dim]"
+
+def _size_and_bytes(pytree: Any) -> Tuple[int, int]:
+  size = sum(
+    (array.size
+    for array in jax.tree_leaves(pytree)),
+    0
+  )
+  # number of bytes
+  bytes = sum(
+    (array.size * array.dtype.itemsize
+    for array in jax.tree_leaves(pytree)),
+    0
+  )
+  return size, bytes
```
Note: `bytes` is a reserved keyword in Python.

```suggestion
  leaves = jax.tree_leaves(pytree)
  size = sum(x.size for x in leaves)
  num_bytes = sum(x.size * x.dtype.itemsize for x in leaves)
```

cgarciae replied:
Changed to `_bytes`

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+
+def _as_yaml_str(value) -> str:
+  
+  if hasattr(value, "__len__") and len(value) == 0:
```
Isn't this more general?

```suggestion
  if not value:
```

cgarciae replied:
If `value` is something like `0` or `None` it would give the wrong output.

marcvanzee replied:
Make sense! Let's keep it as is then.

marcvanzee commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+  )
+  return file.getvalue().replace("\n...", "").replace("'", "").strip()
+
+def _format_value(value):
+  if hasattr(value, "shape") and hasattr(value, "dtype"):
+    shape_str = ",".join(map(str, value.shape))
+    return f"[dim]{value.dtype}[/dim][{shape_str}]"
+  else:
+    return str(value)
```
How about this? It will output shapes as `(3, 4)` instead of `[3, 4]`, but that is actually quite standard. Or did you have a reason why you want to use square brackets?

```suggestion
def _format_value(value):
  if isinstance(value, (np.ndarray, jnp.ndarray)):
    return f"[dim]{value.dtype}[/dim][{value.shape}]"
  else:
    return str(value)
```

cgarciae replied:
I'll be keeping
```python
if hasattr(value, "shape") and hasattr(value, "dtype"):
```
as `ShapeDtypeStruct` created by `eval_shape` is not detected as an instance of `jnp.ndarray`.

jheek replied:
A square bracket shape is quite standard in rendering I think. See for example how jaxpr's are rendered when you print the output of jax.make_jaxpr. I think the intuition borrows from the fact that you always index a multi-dimensional array with square brackets but instead of the index you now put the length.

marcvanzee reviewed (changes requested):
Some overall comments:

* The build is failing in this PR, but the error looks unrelated. Maybe just sync to main will resolve it?
* We use single quotes in the Flax codebase, so please change that everywhere.

cgarciae commented on flax/linen/module.py:
```diff
@@ -1353,6 +1355,95 @@ def __call__(self, x):
+      A string summarizing the Module.
+    """
+
+    tabulate_fn = tabulate(self, rngs, method=method, mutable=mutable, 
```
Sound good.

cgarciae commented on flax/linen/summary.py:
```diff
@@ -0,0 +1,431 @@
+  """
+  path: Tuple[str, ...]
+  outputs: Any
+  module_variables: Dict[str, Dict[str, Array]] # variables of the row's corresponding module
+  counted_variables: Dict[str, Dict[str, Array]] # all variables that should be counted for this row
```
Makes sense! I think they where created before the docs strings just so I knew what they did 😅

cgarciae commented:
Hey @marcvanzee! Thanks for detailed review.

I think the only thing missing is how do we want to render arrays:

With this change values will render like this:

**original**
```
bias: float32[2]
kernel: float32[4,2]
```
**new proposal**
```
bias: float32[(2,)]
kernel: float32[(4, 2)]
```
**alternative proposal**
```
bias: float32(2)
kernel: float32(4, 2)
```

marcvanzee commented:
> Hey @marcvanzee! Thanks for detailed review.
> 
> I think the only thing missing is how do we want to render arrays:
> 
> **original**
> 
> ```
> bias: float32[2]
> kernel: float32[4,2]
> ```
> 
> **new proposal**
> 
> ```
> bias: float32[(2,)]
> kernel: float32[(4, 2)]
> ```
> 
> **alternative proposal**
> 
> ```
> bias: float32(2)
> kernel: float32(4, 2)
> ```

Having read @jheek , I think the original was actually fine, so feel free to keep using that!

cgarciae commented:
@marcvanzee reverted back to square brackets `[...]`. Only one comment pending for resolution.

marcvanzee reviewed (approved):
Looks good to me!

Minor comment for the next time: It wasn't entirely clear to me whether I should review the PR again or whether you were still addressing my comments. You can explicitly re-request a review in the main tab of the PR by clicking the cycled arrows next to a reviewer's name. This is quite helpful since it can happen that you are already addressing some of my comments, but you do not yet want me to do the review again. So then it helps if we both know that I won't do another review until you've explicitly requested it.

jheek reviewed (approved):
Looks great!

marcvanzee commented:
@cgarciae this is currently blocked on internal tests failing. We should just make sure we link in `rich` and get access to it, I'll look into it.

marcvanzee commented:
@cgarciae this change seems to add back all deprecated HOWTOs. perhaps something went wrong when syncing to main, could you please revert those changes?

marcvanzee commented:
@cgarciae Update: I will update it in the internal version so no action needed.

cgarciae commented:
Ok, or should I just sync again with main?

cgarciae commented:
@marcvanzee I cleaned the branch, don't know why it caught these other changes.

marcvanzee commented:
This change is committed but the PR is still open: https://github.com/google/flax/commit/ef6bf4054c30271a58bfabb58f3d0049ef5d851a

I am closing this PR, thank you @cgarciae!!! Awesome change!

trvs2cool commented:
This works really nicely for Modules using the call method.  

However, for modules using other method, it does not work.  The issue is that module.init with capture_intermediates=True only captures [__call__](https://github.com/google/flax/blob/8a148dbd5de4c327dd6b7c931a8b649431e8274e/flax/linen/module.py#L1145) intermediates (See [summary.py:202](https://github.com/google/flax/blob/8a148dbd5de4c327dd6b7c931a8b649431e8274e/flax/linen/summary.py#L202)).

I think the fix is fairly simple.  In [summary.py:184](https://github.com/google/flax/blob/8a148dbd5de4c327dd6b7c931a8b649431e8274e/flax/linen/summary.py#L184)) just pass the output_methods into the capture_intermediates by a lambda:

```  
    capture_intermediates_fn = lambda x,y: y in output_methods or y == '__call__'

    shape_variables = jax.eval_shape(lambda: module.init(
        rngs,
        *args,
        method=method,
        mutable=mutable,
        capture_intermediates=capture_intermediates_fn,
        **kwargs,
    ))```

marcvanzee commented:
@trvs2cool thanks for spotting this! Do you mind creating a separate issue for this problem?
